### ISSUE TITLE
Psalm is complaining that a variable is a value that it can never be

### ISSUE THREAD

**Issue by ollieread** [vimeo/psalm#9657]
I've got Psalm 5.9.0 running on a project with error level 1, but I'm getting errors about things that aren't issues. 

I have the following method:
```php
/**
 * @template R of mixed
 *
 * @param \Smpl\Functional\Contracts\Func<V, R>|\Smpl\Functional\Contracts\BiFunc<K, V, R> $mapper
 *
 * @return \Smpl\Collections\Contracts\Stream<K, R, C>
 *
 * @psalm-this-out self<K, R, C>
 */
public function map(BiFunc|Func $mapper): Contracts\Stream
{
    $elements = [];

    /**
     * @var K $key
     * @var V $value
     */
    foreach ($this->collection as $key => $value) {
        if ($mapper instanceof BiFunc) {
            /** @var R $value */
            $value = $mapper->apply($key, $value);
        } else {
            $value = $mapper->apply($value);
        }

        $elements[$key] = $value;
    }

    $this->collection = $this->collection->fresh($elements);

    return $this;
}
```

I'm getting the following errors for the following lines.

```php
$value = $mapper->apply($key, $value);
```

This produces the following error, complaining that the type of `$mapper` is `Func`, despite it being wrapped in an `if` statement that makes sure it isn't.

```
ERROR: InvalidArgument - src/Collections/Stream.php:118:41 - Argument 1 of Smpl\Functional\Contracts\Func::apply expects V:Smpl\Collections\Stream as mixed, but K:Smpl\Collections\Stream as mixed provided (see https://psalm.dev/004)
                $value = $mapper->apply($key, $value);
```

Amusingly, I'm also getting:

```
ERROR: TooFewArguments - src/Collections/Stream.php:120:35 - Too few arguments for Smpl\Functional\Contracts\BiFunc::apply - expecting arg2 to be passed (see https://psalm.dev/025)
                $value = $mapper->apply($value);

```

For the following line that exists in the `else` branch, and will literally never be touched if `$mapper` is in fact an instance of `BiFunc`:

```php
$value = $mapper->apply($value);
```

**Comment by orklah** [comment]
Please try to reproduce on psalm.dev

It's hard to think about why it happens without knowing for example, the hierarchy between Func and BiFunc

**Comment by ollieread** [comment]
> Please try to reproduce on psalm.dev
> 
> It's hard to think about why it happens without knowing for example, the hierarchy between Func and BiFunc

I tried, but I can't recreate it. `Func` and `BiFunc` have no hierarchy. They are `Func<mixed, mixed>` and `BiFunc<mixed, mixed, mixed>`.

**Comment by ollieread** [comment]
If I change the whole `if` statement to the following:
```php
if ($mapper instanceof BiFunc) {
    $value = $mapper->apply($key, $value);
} else if ($mapper instanceof Func) {
    $value = $mapper->apply($value);
}
```
It still complains. In fact, the only way it doesn't, is if I make sure the second call **ISN'T** a `BiFunc`, which is pretty ridiculous,

```php
if ($mapper instanceof BiFunc) {
    $value = $mapper->apply($key, $value);
} else if ($mapper instanceof Func && ! ($mapper instanceof BiFunc)) {
    $value = $mapper->apply($value);
}
```

**Comment by ollieread** [comment]
The two interfaces `Func` and `BiFunc` can be seen here: https://gist.github.com/ollieread/97f46aa77d85c5006f51244845a4d100

**Comment by ygottschalk** [comment]
Somewhat of a reproducer: https://psalm.dev/r/07b8322f27

What (I think what) happens here: Psalm thinks `$mapper` might implement `Func` and `BiFunc` at the same time. That is why psalm can not pin down the type of `$mapper->apply` just with the one `instanceof`-check, bcs psalm thinks it still might implement both function contracts.

On the other hand it should be possible to know that in the else case `$mapper` can  not be of instance `BiFunc`.  Possibly some bug with type reconciliation and template params I'd guess...

**Comment by ollieread** [comment]
> Somewhat of a reproducer: https://psalm.dev/r/07b8322f27
> 
> 
> 
> What (I think what) happens here: Psalm thinks `$mapper` might implement `Func` and `BiFunc` at the same time. That is why psalm can not pin down the type of `$mapper->apply` just with the one `instanceof`-check, bcs psalm thinks it still might implement both function contracts.
> 
> 
> 
> On the other hand it should be possible to know that in the else case `$mapper` can  not be of instance `BiFunc`.  Possibly some bug with type reconciliation and template params I'd guess...

I think you might be right, in terms of what is happening, because getting errors that it is simultaneously both implementations is...weird.

**Comment by orklah** [comment]
If we try to display $mapper type: https://psalm.dev/r/882d7306e7

It outputs this (I cleaned up the output for easier reading):
`$mapper: BiFunc|(BiFunc&Func<V, R>)|(BiFunc<K, V, R>&Func<V, R>)|(Func<V, R>&Func<V, R>)`

That seems a little odd to me, BiFunc seems to have lost its templates in some cases and we have an intersection with Func in both sides.

So I guess we have an issue in here. But once that is fixed, we should have something like 

`BiFunc<K, V, R>|Func<V, R>|(BiFunc<K, V, R>&Func<V, R>)`

Checking if we have an instance of BiFunc does not allow to exclude that the variable is also a Func and I don't think Psalm allows to document that an interface is exclusive
